include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp

```diff
@@ -249,25 +249,101 @@ template<int dim, typename Number>
 void
 SpatialOperatorBase<dim, Number>::distribute_dofs()
 {
+  if(param.spatial_discretization == SpatialDiscretization::L2)
+  {
+    fe_u = std::make_shared<dealii::FESystem<dim>>(dealii::FE_DGQ<dim>(param.degree_u), dim);
+  }
+  else if(param.spatial_discretization == SpatialDiscretization::HDIV)
+  {
+    // The constructor of FE_RaviartThomas takes the degree in tangential direction as an argument.
+    fe_u = std::make_shared<dealii::FE_RaviartThomasNodal<dim>>(param.degree_u - 1);
+  }
+  else
+    AssertThrow(false, dealii::ExcMessage("FE not implemented."));
+
   // enumerate degrees of freedom
   dof_handler_u.distribute_dofs(*fe_u);
   dof_handler_p.distribute_dofs(fe_p);
   dof_handler_u_scalar.distribute_dofs(fe_u_scalar);
 
-  unsigned int const ndofs_per_cell_velocity =
-    dealii::Utilities::pow(param.degree_u + 1, dim) * dim;
+  if(param.spatial_discretization == SpatialDiscretization::HDIV)
+  {
+    // We need to make sure the normal dofs are shared between cells on the periodic boundaries,
+    // since these are continuous for HDIV.
+    dealii::IndexSet relevant_dofs;
+    dealii::DoFTools::extract_locally_relevant_dofs(dof_handler_u, relevant_dofs);
+    constraint_u.reinit(relevant_dofs);
+
+    auto const & periodic_face_pair = grid->periodic_faces;
+    for(auto const & face : periodic_face_pair)
+      dealii::DoFTools::make_periodicity_constraints(
+        dof_handler_u,
+        face.cell[0]->face(face.face_idx[0])->boundary_id(),
+        face.cell[1]->face(face.face_idx[1])->boundary_id(),
+        face.face_idx[0] / 2,
+        constraint_u);
+
+
+    // Set boundaries constraints
+    if(boundary_descriptor->velocity->symmetry_bc.empty() == false)
+    {
+      for(auto bc : boundary_descriptor->velocity->symmetry_bc)
+        dealii::VectorTools::project_boundary_values_div_conforming(
+          dof_handler_u, 0, *(bc.second), bc.first, constraint_u, *get_mapping());
+    }
+    if(boundary_descriptor->velocity->dirichlet_bc.empty() == false)
+    {
+      AssertThrow(
+        false,
+        dealii::ExcMessage(
+          "Dirichlet BC are not properly implemented for HDIV. The BC needs to be strongly applied in order to obtain an exactly divergence-free solution."));
```

Comment:
Dirichlet BC are -> Dirichlet BCs are

Is it the whole boundary condition that is applied strongly, or just the normal component that is applied strongly?

Reply:
I'm not sure about this. I think that one still needs to take the piola transform into account. Maybe the use of `VectorTools::project_boundary_values_div_conforming` is enough here?

Reply:
@kronbichler what do you think?

Reply:
The project function in deal.II will only affect the normal component of the velocity, tangential components need to be handled weakly just as is the case at internal faces. I would think that `project_boundary_values_div_conforming` picks up the right normal direction (including Piola), but I guess that is a topic to be investigated more in the future.

Reply:
For now, we should adjust the comment and write that the normal component of the velocity needs to be applied strongly.

Reply:
Done here https://github.com/exadg/exadg/pull/233/commits/54b505e18c5f6c9e4fa220627c351cbab07330f3 !